### PR TITLE
snapstate: check gadget model constraints in checkSnap

### PR DESCRIPTION
**Pull request by mvo5**
The current code that checks snaps on refresh does not check
the model constraints. This commit changes that so that no
gadgets can be installed that violate the model constraints.

One more check in firstboot is needed in a followup.

### CODE REVIEW

File: gadget/gadget.go [hunk 1]
```diff
@@ -375,6 +371,24 @@ func ReadInfo(gadgetSnapRootDir string, constraints *ModelConstraints) (*Info, e
...
+// root directory. If constraints is nil, ReadInfo will just check for
+// self-consistency, otherwise rules for the classic or system seed cases are
+// enforced.
+func ReadInfo(gadgetSnapRootDir string, constraints *ModelConstraints) (*Info, error) {
+	gadgetYamlFn := filepath.Join(gadgetSnapRootDir, "meta", "gadget.yaml")
+	return readInfo(ioutil.ReadFile, gadgetYamlFn, constraints)
+}
+
+// ReadInfoFromSnapFile reads the gadget specific metadata from
+// meta/gadget.yaml in the given snapf container. If constraints is
```
Comment (pedronis) [inline gadget/gadget.go]: I would s/snapf/snap/, I know is the argument name, but it just read too much like a typo here (for me)

File: overlord/snapstate/check_snap.go [hunk 2]
```diff
@@ -478,6 +479,22 @@ func checkGadgetOrKernel(st *state.State, snapInfo, curInfo *snap.Info, _ snap.C
...
 
+	// do basic constraints check on the gadget
+	model := deviceCtx.Model()
+	if typ == snap.TypeGadget {
+		validateGadgetModelConstraints := func(model *asserts.Model, snapf snap.Container) error {
+			constraints := &gadget.ModelConstraints{
+				Classic:    model.Classic(),
+				SystemSeed: model.Grade() != asserts.ModelGradeUnset,
+			}
+			_, err = gadget.ReadInfoFromSnapFile(snapf, constraints)
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: this definitely shows that turning ModelConstraints into an interface subsetting asserts.Model is probably a cleaner path forward

File: overlord/snapstate/check_snap.go [hunk 3]
```diff
@@ -478,6 +479,22 @@ func checkGadgetOrKernel(st *state.State, snapInfo, curInfo *snap.Info, _ snap.C
 		return nil
 	}
 
+	// do basic constraints check on the gadget
+	model := deviceCtx.Model()
+	if typ == snap.TypeGadget {
+		validateGadgetModelConstraints := func(model *asserts.Model, snapf snap.Container) error {
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: I'm not sure why the local function?

Comment (mvo5) [reply overlord/snapstate/check_snap.go]: Silly me, this is from a previous iteration. I fixed this now.

File: overlord/snapstate/check_snap.go (hunk 2 above)
Comment (mvo5) [reply overlord/snapstate/check_snap.go]: I will explore this now.

**Comment by mvo5** [comment]
Samuele is looking into an alternative idea, marking as blocked until he finishes this.

**Comment by mvo5** [comment]
No longer needed with https://github.com/snapcore/snapd/pull/7892

**Pull request closed without merging**